src/test/concurrency/transaction_manager_test.cpp

```diff
@@ -0,0 +1,88 @@
+  auto& manager = TransactionManager::get();
+
+  EXPECT_EQ(get_active_snapshot_commit_ids().size(), 0);
+  EXPECT_EQ(manager.get_lowest_active_snapshot_commit_id(), MvccData::MAX_COMMIT_ID);
+
+  const auto t1_context = manager.new_transaction_context();
+  const auto t2_context = manager.new_transaction_context();
+  const auto t3_context = manager.new_transaction_context();
+
+  const CommitID t1_snapshot_commit_id = t1_context->snapshot_commit_id();
+  const CommitID t2_snapshot_commit_id = t2_context->snapshot_commit_id();
+  const CommitID t3_snapshot_commit_id = t3_context->snapshot_commit_id();
+  const auto vec = std::vector<CommitID>{t1_snapshot_commit_id, t2_snapshot_commit_id, t3_snapshot_commit_id};
+
+  EXPECT_EQ(get_active_snapshot_commit_ids().size(), 3);
+  EXPECT_TRUE(std::find(get_active_snapshot_commit_ids().cbegin(), get_active_snapshot_commit_ids().cend(),
+                        t1_snapshot_commit_id) != get_active_snapshot_commit_ids().cend());
+  EXPECT_TRUE(std::find(get_active_snapshot_commit_ids().cbegin(), get_active_snapshot_commit_ids().cend(),
+                        t2_snapshot_commit_id) != get_active_snapshot_commit_ids().cend());
+  EXPECT_TRUE(std::find(get_active_snapshot_commit_ids().cbegin(), get_active_snapshot_commit_ids().cend(),
+                        t3_snapshot_commit_id) != get_active_snapshot_commit_ids().cend());
+  EXPECT_EQ(manager.get_lowest_active_snapshot_commit_id(), *std::min_element(vec.cbegin(), vec.cend()));
+
+  t1_context->commit();
+  deregister_transaction(t1_context->snapshot_commit_id());
+
+  EXPECT_EQ(get_active_snapshot_commit_ids().size(), 2);
+  EXPECT_TRUE(std::find(get_active_snapshot_commit_ids().cbegin(), get_active_snapshot_commit_ids().cend(),
+                        t1_context->snapshot_commit_id()) != get_active_snapshot_commit_ids().cend());
+  EXPECT_TRUE(std::find(get_active_snapshot_commit_ids().cbegin(), get_active_snapshot_commit_ids().cend(),
+                        t3_context->snapshot_commit_id()) != get_active_snapshot_commit_ids().cend());
+  EXPECT_EQ(manager.get_lowest_active_snapshot_commit_id(), t2_context->snapshot_commit_id());
+
+  t3_context->commit();
+  deregister_transaction(t3_context->snapshot_commit_id());
+
+  EXPECT_EQ(get_active_snapshot_commit_ids().size(), 1);
+  EXPECT_TRUE(std::find(get_active_snapshot_commit_ids().cbegin(), get_active_snapshot_commit_ids().cend(),
+                        t2_context->snapshot_commit_id()) != get_active_snapshot_commit_ids().cend());
+  EXPECT_EQ(manager.get_lowest_active_snapshot_commit_id(), t2_context->snapshot_commit_id());
+
+  t2_context->commit();
+  deregister_transaction(t2_context->snapshot_commit_id());
+
+  EXPECT_EQ(get_active_snapshot_commit_ids().size(), 0);
+  EXPECT_EQ(manager.get_lowest_active_snapshot_commit_id(), MvccData::MAX_COMMIT_ID);
+
+  // To prevent exceptions in TransactionContext destructor
+  register_transaction(t1_snapshot_commit_id);
```

Comment:
Shouldn't it actually be `TransactionContext::commit` and `TransactionContext::rollback` that end the lifecycle of a transaction and de-register the transaction from the TransactionManager? This would reduce the lifetime of the Transaction, which is imo a good thing, and render void the necessity to have these hacky `register_transaction()` calls here.

Reply:
We've had a discussion regarding this a few weeks back: To keep `TransactionManager`'s `_active_snapshot_commit_ids` consistent, we need to make sure that `deregister_transaction` is called only once per `TransactionContext`. Since `commit` and `abort` can be called multiple times we would need a bool flag  and additional checks to ensure `exactly-once` semantics. 
When using the destructor of `TransactionContext` to call `deregister_transaction`, we do not need any checks. In addition, the code stays clean and compact. 
The delayed call of `deregister_transaction`is feasible.

Remember, the hacky use of `register_transaction` here is required for this test only since it breaks with conventions. `TransactionManager` is usually the only class calling `register_transaction`.

Reply:
> Since `commit` and `abort` can be called multiple times

This is news to me. Also, this seems to suggest otherwise: 
https://github.com/hyrise/hyrise/blob/master/src/lib/concurrency/transaction_context.hpp#L133

Reply:
Thanks for the hint @mrks, you are right! In consequence, there is one argument left for the current implementation: Minimal use of code – one line of code in the destructor vs. multiple spread across `TransactionContext`.

So, should we switch again, to the approach @mrzzzrm suggested?

Reply:
I think both have their advantages and disadvantages. Once we work more with modifying transactions, we will see if there is a material difference between the two. We might also see questions like "What happens to contexts of lost network connections" and alike.

For now, I'd leave it the way it is. In order to slowly reduce the number of open PRs, I wouldn't see this as a blocker.